drafts/Asynchronous-programming-made-resilient-with-MPFT.adoc

```diff
@@ -0,0 +1,230 @@
+
+But there's one in particular which is going to help you make your asynchronous methods resilient to faults: https://download.eclipse.org/microprofile/microprofile-fault-tolerance-2.1/apidocs/org/eclipse/microprofile/faulttolerance/Asynchronous.html[@Asynchronous].
+
+Let's show you what `@Asynchronous` does and how it does it.
+
+== What does `@Asynchronous` do?
+The `@Asynchronous` annotation makes your method run asynchronously, meaning that rather than running straightaway on the main thread when you call it, it will run some time later, usually on another thread. To do this, we need to add the `@Asynchronous` annotation to the method, and we change the method to return either a https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html[CompletionStage] (preferable) or a https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html[Future]. In our method body, we can use `CompletableFuture.completedFuture()` to make our result compatible with the new return type.
+
+[source,java]
+----
+@Asynchronous
+public CompletionStage<String> serviceA() { 
+
+	return CompletableFuture.completedFuture("serviceA");
+
+}
+----
+
+What does this do when we call the method?
+
+Well, usually when you call a method, you call it, and then it returns a result. However, when we use the `@Asynchronous` annotation, Fault Tolerance gets in between. When you call the method, Fault Tolerance first schedules your method to run later on a different thread, then it returns you a `CompletionStage` or `Future`. This isn't the `CompletionStage` or `Future` that's returned from your method, because your method hasn't run yet, it's an object created entirely by Fault Tolerance. 
+
+image::/img/blog/FT-basic-asynchronous-execution.png[Fault Tolerance asynchronous execution flow]
+
+Sometime later, your method runs and returns a `CompletionStage` or `Future`. Now, Fault Tolerance will go and complete the `CompletionStage` or `Future` object which was returned to the caller earlier, with the value that was returned from your method. You might be wondering *_"What if the returned result is an exception? What can I do?"_*.  Read along.
+
+== How you can use `@Asynchronous`
+So `@Asynchronous` sounds great, right? But how are we going to use it to make our asynchronous programs resilient? Let's go through 2 use-cases:
+
+1. Applying Fault Tolerance to an API which is already asynchronous
+2. Using `@Asynchronous` annotation to do two things in parallel with resilience
+
+=== 1. Applying to An Already Asynchronous API
+Another use is to apply Fault Tolerance to an API which is already asynchronous. We can do this with any API which does something Asynchronously and returns a `CompletionStage`.
+
+In this example, we'll use the JAX-RS client API for calling remote REST services, introduced in JAX-RS 2.1. We can build up a request to fetch a String from a given URL with a GET request and the return type is a `CompletionStage` of String:
```

Comment:
These first two parts don't join together very well.

The intro says `@Asynchronous` makes your method run asynchronously. Then the blog goes straight into using `@Asynchronous` with an already asynchronous API so that fault tolerance is applied around the completion stage result.

There's nothing that links these two parts, so it feels like you're leading the reader along one path, then you jump over to something completely different.

I'm not sure of the best way to fix that. I think the problem is that if you talk about what `@Asynchronous` does, the bigger part is that it runs things asynchronously, and the user needs to be aware of that. However, in this blog, you wanted to lead with applying FT to an asynchronous API.

Reply:
'Another use' sounds out of place as it is the first use. You can update to" One usage is to apply ...".

Based on what @Azquelt said, I think if you rephrase the first one to focus on "Applying `@Asynchronous` to An Already Asynchronous method". By the way, it should not be Asynchronous API, it should be Asynchronous Method.

Reply:
Ah good spot Emily. I did not notice it still says "Another use"

I'll include a potential solution in the next draft (Draft 7) and we can see if it resolves this issue.

Reply:
What was wrong with Asynchronous API?

"Applying Fault Tolerance to an Already Asynchronous Method" sounds a bit like you're applying fault tolerance to a method which is already annotated with `@Asynchronous`.

Reply:
"Applying FT to an already asynchronous method invocation"